Mempool util: Add RBF diagram checks for single chunks against clusters of size 2

instagibbs opened the pull request:
This is a smaller piece of https://github.com/bitcoin/bitcoin/pull/28984 broken off for easier review.

Up to date explanation of diagram checks are here: https://delvingbitcoin.org/t/mempool-incentive-compatibility/553

This infrastructure has two near term applications prior to cluster mempool:
1) Limited Package RBF(https://github.com/bitcoin/bitcoin/pull/28984): We want to allow package RBF only when we know it improves the mempool. This narrowly scoped functionality allows use with v3-like topologies, and will be expanded at some point post-cluster mempool when diagram checks can be done efficiently against bounded cluster sizes.
2) Replacement for single tx RBF(in a cluster size of up to two) against conflicts of up to cluster size two. `ImprovesFeerateDiagram` interface will have to change for this use-case, which is a future direction to solve certain pins and improve mempool incentive compatibility: https://delvingbitcoin.org/t/ephemeral-anchors-and-mev/383#diagram-checks-fix-this-3

And longer-term, this would be the proposed way we would compute incentive compatibility for all conflicts, post-cluster mempool.

instagibbs committed:
Add FeeFrac utils

Co-authored-by: Suhas Daftuar <sdaftuar@chaincode.com>
Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

instagibbs committed:
Add FeeFrac unit tests

Co-authored-by: Suhas Daftuar <sdaftuar@chaincode.com>

instagibbs committed:
fuzz: Add fuzz target for ImprovesFeerateDiagram

Co-authored-by: Suhas Daftuar <sdaftuar@chaincode.com>

instagibbs committed:
test: Add tests for CompareFeerateDiagram and CheckConflictTopology

instagibbs committed:
fuzz: fuzz diagram creation and comparison

Co-authored-by: Suhas Daftuar <sdaftuar@chaincode.com>
Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

instagibbs commented:
ready for review

cc @glozow @ismaelsadeeq @achow101

instagibbs commented on src/test/feefrac_tests.cpp:
```diff
@@ -0,0 +1,126 @@
+    FeeFrac p1{1000, 100}, p2{500, 300};
+    FeeFrac sum{1500, 400};
+    FeeFrac diff{500, -200};
+    FeeFrac empty{0, 0};
```
Even better, I'll add a test that they're the same

instagibbs replied:
done

instagibbs commented on src/txmempool.cpp:
```diff
@@ -1248,3 +1249,121 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+
+    std::vector<FeeFrac> new_chunks;
+
+    // Step 2: build the NEW = OLD - CON + CNK diagram
+
+    // OLD - CON: Any parents of direct conflicts that are not conflicted themselves
+    for (auto direct_conflict : direct_conflicts) {
+        // If a direct conflict has an ancestor that is not in all_conflicts,
+        // it can be affected by the replacement of the child.
+        if (direct_conflict->GetMemPoolParentsConst().size() > 0) {
+            // Grab the parent.
+            const CTxMemPoolEntry& parent = direct_conflict->GetMemPoolParentsConst().begin()->get();
+            if (!all_conflicts.count(mapTx.iterator_to(parent))) {
+                // This transaction would be left over, so add to the new
+                // diagram.
+                new_chunks.emplace_back(parent.GetModifiedFee(), parent.GetTxSize());
+            }
+        }
+    }
+    // + CNK
+    new_chunks.emplace_back(replacement_fees, int32_t(replacement_vsize));
```
These are shorthand labels from https://delvingbitcoin.org/t/post-clustermempool-package-rbf-per-chunk-processing/190 which helped me think through preciesely how it maps onto the concepts laid out.

I'd like to keep track of these ideas somehow, maybe by defining the labels?

ismaelsadeeq replied:
Yeah, maybe to be consistent define the labels and do the same in the  `OLD` `chunk` computation loop above.
Though  the explicit comments in the `OLD`  `chunk` computation loop is more easier to parse for me :).

instagibbs replied:
took some of hte language, and explicitly defined each term

instagibbs commented on src/test/feefrac_tests.cpp:
```diff
@@ -0,0 +1,126 @@
+{
+    FeeFrac p1{1000, 100}, p2{500, 300};
+    FeeFrac sum{1500, 400};
+    FeeFrac diff{500, -200};
```
this is testing that subtraction works correctly, not that an individual chunk can have negative size

ismaelsadeeq replied:
Yes, My question was... whats the rationale for `FeeFrac` to be generally have negative size ?

instagibbs replied:
We use subtraction, and should try to handle that?

sdaftuar replied:
> If `old`, `new`, and `tail` are CFeeFrac objects, this condition is exactly `!((new - old) << tail)`. If new is smaller than old, the `new - old` object has negative size.

@sipa With the various `Assume()` calls that check that if the size is 0, the fee must also be zero, doesn't that mean that we can't really write code like this example you gave?  Unless we checked that new != old first -- otherwise you might create a FeeFrac with 0 size and non-zero fee.

sdaftuar replied:
I think we should drop the `Assume()` calls, as otherwise I think we ought to do more to ensure that using operator- on FeeFrac's is safe everywhere that might be invoked.

ismaelsadeeq commented on src/policy/rbf.cpp:
```diff
@@ -181,3 +201,108 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
+
+    const auto err_string{pool.CalculateFeerateDiagramsForRBF(replacement_fees, replacement_vsize, direct_conflicts, all_conflicts, old_diagram, new_diagram)};
+
+    if (err_string.has_value()) {
+        return strprintf("unable to compute mining score");
+    }
+
+    if (!CompareFeerateDiagram(old_diagram, new_diagram)) {
+        return strprintf("insufficient feerate");
+    }
```
I meant what is being returned by `err_string` and `CompareFeerateDiagram` was not used. why were you unable to compute mining score, same for `CompareFeerateDiagram` why is the fee rate insufficient.

instagibbs replied:
CompareFeerateDiagram won't return much very interesting, but you're right for the `err_string`, will return it

instagibbs replied:
done

instagibbs committed:
Add fuzz test for FeeFrac

instagibbs committed:
test: unit test for ImprovesFeerateDiagram

instagibbs commented on src/txmempool.cpp:
```diff
@@ -1248,3 +1249,121 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+        if (txiter->GetCountWithAncestors() > 1) {
+            // We'll add chunks for either the ancestor by itself and this tx
+            // by itself, or for a combined package.
+            FeeFrac package{txiter->GetModFeesWithAncestors(), int32_t(txiter->GetSizeWithAncestors())};
```
done

instagibbs commented on src/txmempool.cpp:
```diff
@@ -1248,3 +1249,121 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+    // direct_conflicts that is at its own fee/size, along with the replacement
+    // tx/package at its own fee/size
+
+    // old diagram will consist of each element of direct_conflicts either at
```
done

instagibbs commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,158 @@
+    }
+};
+
+void BuildDiagramFromUnsortedChunks(std::vector<FeeFrac>& chunks, std::vector<FeeFrac>& diagram);
```
added comment

sdaftuar replied:
> The correctness of the result of `BuildDiagramFromUnsortedChunks` very much depends on the assumption that the chunks can be reordered w.r.t. one another, which seems to be the case in the current PR, but will very much not be the case going forward in a general cluster mempool world.

No objection to your suggested change here, but I don't follow this comment -- in a general cluster mempool world, unless we were to bound chunk sizes to something smaller than a cluster size (resulting in the possibility of chunk feerates not being monotonically decreasing), I think for the purposes of a feerate diagram it is fine to sort by chunk feerate and permit equal-feerate-chunks to be reordered?  From the perspective of a feerate diagram, two such diagrams are equivalent.

If you're suggesting that we want to preserve the ability in the future to impose a chunk size limit, then yes I agree that we should rewrite this...  Is there another case that I'm overlooking?

instagibbs replied:
@sdaftuar that matches my understanding.

>Removing the sorting step in this function (moving it to the caller). The function can then take Span<const FeeFrac> as input too, and perhaps in a further iteration (in a later PR) this function could then be turned into one that actually performs chunking too. 

Took this suggestion

instagibbs commented on src/util/feefrac.cpp:
```diff
@@ -0,0 +1,21 @@
+
+void BuildDiagramFromUnsortedChunks(std::vector<FeeFrac>& chunks, std::vector<FeeFrac>& diagram)
+{
+    diagram.clear();
```
done

instagibbs commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,158 @@
+        size += other.size;
+    }
+
+    /** Subtrack size and size of another FeeFrac from this one. */
```
done

instagibbs committed:
Implement ImprovesFeerateDiagram

This new function takes the populated sets of
direct and all conflicts computed in the current
mempool, assuming the replacements are a single
chunk, and computes a diagram check.

The diagram check only works against cluster
sizes of 2 or less, and fails if it encounters
a different topology.

Co-authored-by: Suhas Daftuar <sdaftuar@chaincode.com>

instagibbs commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    /** Check if a FeeFrac object is worse than another. */
+    friend inline bool operator<(const FeeFrac& a, const FeeFrac& b) noexcept
```
done

instagibbs commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    /** Check if two FeeFrac objects are different (either fee or size differs). */
+    friend inline bool operator!=(const FeeFrac& a, const FeeFrac& b) noexcept
```
done

instagibbs commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,202 @@
+#include <vector>
+#include <util/check.h>
+
+namespace {
```
done

instagibbs commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -0,0 +1,126 @@
+    int sign_b = (b1 == 0 ? 0 : b1 < 0 ? -1 : 1) * (b2 == 0 ? 0 : b2 < 0 ? -1 : 1);
+    if (sign_a != sign_b) return sign_a <=> sign_b;
+
+    // Compute absolute values.
```
done

instagibbs commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -0,0 +1,126 @@
+    std::array<uint32_t, 4> ret{0, 0, 0, 0};
+
+    /** Perform ret += v << (32 * pos), at 128-bit precision. */
+    auto add_fn = [&](uint64_t v, int pos) {
```
looks good, done

instagibbs commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -0,0 +1,126 @@
+        assert(carry == 0);
+    };
+
+    // Multiply the 4 individual limbs (schoolbook multiply, with base 2^64).
```
done

instagibbs commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -0,0 +1,126 @@
+/* comparison helper for std::vector */
+std::strong_ordering compare_arrays(const std::array<uint32_t, 4>& a, const std::array<uint32_t, 4>& b) {
+    for (size_t i = 0; i < a.size(); ++i) {
+        if (a[i] < b[i]) return std::strong_ordering::less;
```
done

instagibbs commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -0,0 +1,126 @@
+    return ret;
+}
+
+/* comparison helper for std::vector */
```
done

instagibbs commented on src/policy/rbf.h:
```diff
@@ -106,4 +108,23 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
+                                                EXCLUSIVE_LOCKS_REQUIRED(pool.cs);
+
+/** Compares two feerate diagrams. The shorter one is padded with a horizonal line. */
+std::partial_ordering CompareFeerateDiagram(std::vector<FeeFrac>& dia0, std::vector<FeeFrac>& dia1);
```
makes sense, moved

instagibbs commented:
test failure appears to be spurious wallet failure, ready for review

instagibbs commented:
> In https://github.com/sipa/bitcoin/commits/pr29242 I pushed another commit which makes CompareFeerateDiagram not modify the diagrams in-place.

Taken only with minor comment changes, and added another test case or two to cover the iterative nature of the check

murchandamus commented on src/test/rbf_tests.cpp:
```diff
@@ -217,15 +232,113 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
+
+    BOOST_CHECK(std::is_lt(CompareFeerateDiagram(old_diagram, new_diagram)));
+
+    // Feerate of first chunk is sufficiently better, but second chunk is worse.
```
In "test: Add tests for CompareFeerateDiagram and CheckConflictTopology" (b3d415fe84be7edfbed567a79a25d406b438622b):
Nit: I found the comment here confusing. How about: "`new_diagram` is strictly better due to the first chunk, while the second chunk is worse."

instagibbs replied:
tried my own re-explanation. I agree as-is is confusing

murchandamus commented on src/test/rbf_tests.cpp:
```diff
@@ -217,15 +232,113 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
+    BOOST_CHECK(std::is_lt(CompareFeerateDiagram(old_diagram, new_diagram)));
+
+    // Feerate of first chunk is better, but second chunk is worse
+    old_diagram = {FeeFrac{0, 0}, FeeFrac{950, 300}, FeeFrac{1050, 400}};
+    new_diagram = {FeeFrac{0, 0}, FeeFrac{750, 100}, FeeFrac{999, 350}, FeeFrac{1150, 500}};
```
In "test: Add tests for CompareFeerateDiagram and CheckConflictTopology" (b3d415fe84be7edfbed567a79a25d406b438622b):

Nit: in the new diagram, the last two chunks should be one chunk because the last chunk is better (151 sats, 150 vB) than the prior (249 sats, 250 vB).

instagibbs replied:
ah hmm, I'm not sure that matters for correctness of test, but jacked up the last chunk's size just to make it clearer what we're testing

murchandamus commented on src/test/rbf_tests.cpp:
```diff
@@ -217,15 +232,113 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
+
+    // No change in evaluation when tail check needed.
+    BOOST_CHECK(std::is_gt(CompareFeerateDiagram(old_diagram, new_diagram)));
+
+    old_diagram = {FeeFrac{0, 0}, FeeFrac{950, 300}, FeeFrac{1050, 399}};
+
```
In "test: Add tests for CompareFeerateDiagram and CheckConflictTopology" (b3d415fe84be7edfbed567a79a25d406b438622b):

This is the same old_diagram. Did you mean to repeat the same transactions,
```suggestion

    old_diagram = {FeeFrac{0, 0}, FeeFrac{950, 300}, FeeFrac{1050, 399}};
    new_diagram = {FeeFrac{0, 0}, FeeFrac{950, 300}, FeeFrac{1050, 400}};
```
or just test in both directions?
```suggestion
```

instagibbs replied:
I can't recall honestly, removed that line

murchandamus reviewed (commented):
Concept ACK, tentative crACK f094a0370db7085b8a89842de0b6d12272d826cb

instagibbs commented on src/policy/rbf.cpp:
```diff
@@ -19,6 +19,8 @@
 #include <limits>
 #include <vector>
 
+#include <compare>
```
direct_conflicts and all_conflicts. Does that help?

murchandamus replied:
Yeah, that would improve it

instagibbs commented on src/policy/rbf.h:
```diff
@@ -106,4 +115,20 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
+/**
+ * The replacement transaction must improve the feerate diagram of the mempool.
+ * @param[in]   pool                The mempool.
+ * @param[in]   direct_conflicts    Set of txids corresponding to the direct conflicts
+ * @param[in]   all_conflicts       Set of mempool entries corresponding to all conflicts
```
direct_conflicts -> the set of transactions that have at least one input conflicting with a proposed transaction.
all_conflicts -> Everything that would be evicted by the proposed transaction

I'll touch this up a bit but this is the nomenclature elsewhere

instagibbs commented on src/txmempool.h:
```diff
@@ -734,6 +735,23 @@ class CTxMemPool
+     *
+     * @param[in] replacement_fees    Package fees
+     * @param[in] replacement_vsize   Package size
+     * @param[in] direct_conflicts    All transactions that would be removed directly (not via descendants)
```
tried my own explanation

instagibbs commented on src/txmempool.cpp:
```diff
@@ -1248,3 +1249,126 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+                old_chunks.emplace_back(package);
+            } else {
+                // Add two points, one for the parent and one for this child.
+                old_chunks.emplace_back(package-individual);
```
fixed

instagibbs commented on src/txmempool.cpp:
```diff
@@ -1248,3 +1249,126 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+        } else if (descendant_count > 2) {
+            return strprintf("%s has %u descendants, max 1 allowed", txid_string, descendant_count - 1);
+        } else if (has_ancestor && has_descendant) {
+            return strprintf("%s has both ancestor and descendant", txid_string);
```
taken

instagibbs commented on src/txmempool.cpp:
```diff
@@ -1248,3 +1249,126 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+        }
+        // Additionally enforce that:
+        // If we have a parent, we are its only child.
+        // If we have a child,  we are its only parent.
```
done

murchandamus commented on src/txmempool.h:
```diff
@@ -734,6 +735,24 @@ class CTxMemPool
+     * @param[in] replacement_fees    Package fees
+     * @param[in] replacement_vsize   Package size
+     * @param[in] direct_conflicts    All transactions that would be removed directly by
+     *                                having a conflicting input wih a proposed transaction
```
```suggestion
     *                                having a conflicting input with a proposed transaction
```

instagibbs replied:
fixed

murchandamus commented on src/test/rbf_tests.cpp:
```diff
@@ -217,15 +232,154 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
+    BOOST_CHECK(std::is_lt(CompareFeerateDiagram(old_diagram, new_diagram)));
+
+    // Incomparable diagrams
+    old_diagram = {FeeFrac{0, 0}, FeeFrac{950, 300}, FeeFrac{1050, 400}};
+    new_diagram = {FeeFrac{0, 0}, FeeFrac{1000, 300}, FeeFrac{1000, 400}};
```
If all diagrams always start with `FeeFrac{0, 0}`, why do we need to mention that as a first element?

instagibbs replied:
I'll leave this as an ergonomics issue for later debate

murchandamus reviewed (commented):
ACK a14b95129d3a2894b7a41ce919a426bb60f62e35

Note: I am fairly familiar with the concept of feerate diagram checks, but not particularly familiar with the mempool code.

dergoegge commented:
```
$ echo "BAIBAgICAgIABfcN/f11BwAAAPsAAAICAgICAgICAi0AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABwb3J0AgL+//////8hAQICAooBATIBAR4AAgICAgICAgICAgICAgICAgICAgICAgIC/v///////wECAgKKAQEyAQEeAgIiAgICAgICHgACAgIDAgICAgICAiICAgICAExLQAICAgICAgICAv//////////////////////////////////////////////////////AQEBAQEByQEBAQEBAQEB/P8AAC8AGvsAXAEAAgLcAgICAgICAlSbtqZ56uTNNhvmIHMu83ifTNtXiVGR9z3J4We8dnhZegAAAAAAAAD/AAAAAAAHAAAAAAAAXP96AAAAAgICAgAF9w39/QAAAAEAAAABAgICAgICAgICLZCPcnQCAgIC3AICAgICAgJUm7ameerkzRvmIHMu83ifTNtXiVGR9z3J4We8dnhZegAAAAAAAAAAAAAAAAcABgEAAAAAAAAAAAAAAPsCIlwBAAAHAAAAAQAAAAmLXAEVAAAAAAAAAAAAAAAAAAAAXP+B/////////wAAAAAABwAAAAAAAAAAAAAABwAALAACAgIATEtAAgICAgICAgIC//////////////////////////////////////////////////////8BAQEBAQHJAQEBAQEBAQH8/wAALwAa+wBcAQACAtwCAgICAgICVJu2pnnq5M02G+Ygcy7zeJ9M21eJUZH3PcnhZ7x2eFl6AAAAAAAAAP8AAAAAAAcAAAAAAABc/3oAAAACAgICAAX3Df39AAAAAQAAAAECAgICAgICAgItkI9ydAICAgLcAgICAgICAlSbtqZ56uTNG+Ygcy7zeJ9M21eJUZH3PcnhZ7x2eFl6AAAAAAAAAAAAAAAABwAGAQAAAAAAAAAAAAAA+wIiXAEAAAcAAAABAAAACYtcARUAAAAAAAAAAAAAAAAAAABc/4H/////////AAAAAAAHAAAAAAAAAAAAAAAHAAAsAAAAAPsCAlwBAAAHAAAAAQAAAAmLXAEDAwMDAwMD/729vb29vb29vb29vb29vb29vb29vb29vT+9/wUtYmluvb29vb29vb29vb29vb29vb29vb29/wUtYmluZC0U/////wAAAAAAAAAAHAAAAAAAAP///////////////////////4pVyFwAAAD7AgJcAQAABwAAAAEAAAAJi1wBAwMDAwMDA/+9vb29vb29vb29vb29vb29vb29vb29vb0/vf8FLWJpbr29vb29vb29vb29vb29vb29vb29vf8FLWJpbmQtFP////8AAAAAAAAAABwAAAAAAAD///////////////////////+KVchcToownpjPpiYHL06aRtJ04wyb/sdpyYUpxMKugP////////////////////////9t////AQAAAEZFRUVFRUVFRUVFRUVFRUVFRUVFRUVF/////////////////////3r//////////wMDAwP//////w==" | base64 --decode > package_rbf-7d643e62d75f4ac67d3bdeaf9f5fbfd67675806a.crash
$ FUZZ=package_rbf ./src/test/fuzz/fuzz package_rbf-7d643e62d75f4ac67d3bdeaf9f5fbfd67675806a.crash
fuzz_libfuzzer: test/fuzz/rbf.cpp:147: void package_rbf_fuzz_target(FuzzBufferType): Assertion `old_diagram.back().fee - replaced_fee + replacement_fees == new_diagram.back().fee' failed.
```

instagibbs commented:
@dergoegge fuzzer created a circular reference which caused incongruities between descendant counts and actually walking the descendants in mempool. Please run again?

dergoegge commented:
> Please run again?

Running now.

Some coverage reports from the previous runs (one per harness): http://bitcoind-fuzz.dergoegge.de/instagibbs/bitcoin/2024-01-diagram-checks/coverage/

instagibbs commented:
didn't see any fuzzer crashes locally, all comments addressed

instagibbs commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,164 @@
+    /** Fee. */
+    int64_t fee;
+    /** Size. */
+    int32_t size;
```
> If not I think its okay like this.

removed comments, how it's precisely used is an implementation detail outside of this code

instagibbs commented:
rebased

ismaelsadeeq commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,162 @@
+ * The size of a FeeFrac cannot be zero unless the fee is also zero.
+ *
+ * FeeFracs have a total ordering, first by increasing feerate (ratio of fee over size), and then
+ * by decreasing size. The empty FeeFrac (fee and size both 0) sorts last. So for example, the
+ * following FeeFracs are in sorted order:
+ *
+ * - fee=0 size=1 (feerate 0)
+ * - fee=1 size=2 (feerate 0.5)
+ * - fee=2 size=3 (feerate 0.667...)
+ * - fee=2 size=2 (feerate 1)
+ * - fee=1 size=1 (feerate 1)
+ * - fee=3 size=2 (feerate 1.5)
+ * - fee=2 size=1 (feerate 2)
+ * - fee=0 size=0 (undefined feerate)
+ *
```
In implementation they are sorted in increasing fee rate and believe the comments should be ordered like as is? and also the 0 fee and size chuck sorts first. 
```suggestion
 * by decreasing size. The empty FeeFrac (fee and size both 0) sorts first. So for example, the
 * following FeeFracs are in sorted order:
 *
 * - fee=0 size=0 (undefined feerate)
 * - fee=2 size=1 (feerate 2)
 * - fee=3 size=2 (feerate 1.5)
 * - fee=1 size=1 (feerate 1)
 * - fee=2 size=2 (feerate 1)
 * - fee=2 size=3 (feerate 0.667...)
 * - fee=1 size=2 (feerate 0.5)
 * - fee=0 size=1 (feerate 0)
```

instagibbs replied:
I don't understand your comment, sorry. They're listed in increasing feerate, with decreasing size as a tie breaker.

ismaelsadeeq replied:
Yes but in reverse I think.
Suppose to be the other way around no?

A sorted `FeeFracs`  should have undefined chunks first, the highest fee rate chunk second continuously...
If their is a tie, the chunk with lower size comes first.
Hence the are sorted in increasing fee rates, and then by decreasing size.

I've added test to verify this here https://github.com/ismaelsadeeq/bitcoin/commit/8ce89b132f4304a7feda067a88fd0a72330044a6, this test passed on this branch.

ismaelsadeeq replied:
Yes thanks for clarification 👍🏾

ismaelsadeeq replied:
I looked at it the other way around, using the result from `BuildDiagramFromUnsortedChunks` sort which custom > operator @instagibbs.

ismaelsadeeq commented on src/txmempool.cpp:
```diff
@@ -1235,3 +1236,126 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+    return std::nullopt;
+}
+
+std::optional<std::string> CTxMemPool::CalculateFeerateDiagramsForRBF(CAmount replacement_fees, int64_t replacement_vsize, const setEntries& direct_conflicts, const setEntries& all_conflicts, std::vector<FeeFrac>& old_diagram, std::vector<FeeFrac>& new_diagram)
```
Maybe add unit test for `CalculateFeerateDiagramsForRBF` also?

ismaelsadeeq commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -0,0 +1,127 @@
+    FuzzedDataProvider provider(buffer.data(), buffer.size());
+
+    int64_t f1 = provider.ConsumeIntegral<int64_t>();
+    //int64_t f1 = provider.ConsumeIntegralInRange(std::numeric_limits<int64_t>::min() + 1,
+    //                            std::numeric_limits<int64_t>::max() - 1);
+    int32_t s1 = provider.ConsumeIntegral<int32_t>();
+    if (s1 == 0) f1 = 0;
+    FeeFrac fr1(f1, s1);
+    assert(fr1.IsEmpty() == (s1 == 0));
+
+    int64_t f2 = provider.ConsumeIntegral<int64_t>();
+    //int64_t f2 = provider.ConsumeIntegralInRange(std::numeric_limits<int64_t>::min() + 1,
+    //                            std::numeric_limits<int64_t>::max() - 1);
```
Why leave out commend version of f1 and f2.
```suggestion
    int32_t s1 = provider.ConsumeIntegral<int32_t>();
    if (s1 == 0) f1 = 0;
    FeeFrac fr1(f1, s1);
    assert(fr1.IsEmpty() == (s1 == 0));

    int64_t f2 = provider.ConsumeIntegral<int64_t>();
      
```

instagibbs replied:
mistakenly left at some point; removed

ismaelsadeeq reviewed (approved):
Code review ACK 3a2c66cc4f0afad9aff300110b18a39981055d63

I've reviewed the following commits
-  [x] [Add FeeFrac utils `d200d30431368c7637a3490509b9158844b69e0d` ](https://github.com/bitcoin/bitcoin/pull/29242/commits/d200d30431368c7637a3490509b9158844b69e0d)
-  [x]  [Add FeeFrace unit tests `860823fb93212fa78ab928589bd403da462be222`](https://github.com/bitcoin/bitcoin/pull/29242/commits/860823fb93212fa78ab928589bd403da462be222)
- [x] [Implement ImprovesFeerateDiagram `802a104d0e2f810d97567e161e0cf61331052906`](https://github.com/bitcoin/bitcoin/pull/29242/commits/802a104d0e2f810d97567e161e0cf61331052906)
- [x] [test: Add tests for CompareFeerateDiagram and CheckConflictTopology `01f682d49aabec7e07f243a314f699bbca710469`](https://github.com/bitcoin/bitcoin/pull/29242/commits/01f682d49aabec7e07f243a314f699bbca710469)
- [x] [test: unit test for ImprovesFeerateDiagram `3a2c66cc4f0afad9aff300110b18a39981055d63`](https://github.com/bitcoin/bitcoin/pull/29242/commits/3a2c66cc4f0afad9aff300110b18a39981055d63)

And a light review of the fuzzing commits.

The code looks good to me, I did some local testing and mess around with the implementation and unit tests.

I have just a few comments for potential further improvements.

instagibbs committed:
Unit tests for CalculateFeerateDiagramsForRBF

instagibbs commented:
addressed all comments

instagibbs commented on src/test/rbf_tests.cpp:
```diff
@@ -309,6 +309,128 @@ BOOST_FIXTURE_TEST_CASE(improves_feerate, TestChain100Setup)
+    BOOST_CHECK(old_diagram[0] == FeeFrac(0, 0));
+    BOOST_CHECK(old_diagram[1] == FeeFrac(low_fee, low_size));
+    BOOST_CHECK(new_diagram[0] == FeeFrac(0, 0));
+    BOOST_CHECK(new_diagram[1] == FeeFrac(0, 0));
```
this is odd, should we just `Assume()` `replacement_vsize>0` in `CalculateFeerateDiagramsForRBF`?

ismaelsadeeq replied:
Yes, I can't think of a scenario where that will happen!

instagibbs replied:
removed

ismaelsadeeq commented on src/test/rbf_tests.cpp:
```diff
@@ -309,6 +309,128 @@ BOOST_FIXTURE_TEST_CASE(improves_feerate, TestChain100Setup)
+    const CAmount high_fee{CENT};
+
+    // low -> high -> medium fee transactions that would result in two chunks together
+    const auto low_tx= make_tx(/*inputs=*/ {m_coinbase_txns[0]}, /*output_values=*/ {10 * COIN});
```
nit:
```suggestion
    const auto low_tx = make_tx(/*inputs=*/ {m_coinbase_txns[0]}, /*output_values=*/ {10 * COIN});
```

ismaelsadeeq replied:
here an another place below!

ismaelsadeeq commented on src/test/rbf_tests.cpp:
```diff
@@ -309,6 +309,128 @@ BOOST_FIXTURE_TEST_CASE(improves_feerate, TestChain100Setup)
+    BOOST_CHECK(new_diagram[1] == FeeFrac(high_fee, low_size_2));
+    old_diagram.clear();
+    new_diagram.clear();
 }
```
Maybe weshould also check that we can have more than two directly conflicting transactions as long as they are all in a cluster size <=2.

<details>
<summary>diff </summary>

```diff
diff --git a/src/test/rbf_tests.cpp b/src/test/rbf_tests.cpp
index e15fd29da3..561ebd3881 100644
--- a/src/test/rbf_tests.cpp
+++ b/src/test/rbf_tests.cpp
@@ -431,6 +431,53 @@ BOOST_FIXTURE_TEST_CASE(calc_feerate_diagram_rbf, TestChain100Setup)
     BOOST_CHECK(new_diagram[1] == FeeFrac(high_fee, low_size_2));
     old_diagram.clear();
     new_diagram.clear();
+
+    // You can more than two direct conflicts,  if all the directly conflicting transactions are in a cluster size < 2
+    const auto conflict_1 = make_tx(/*inputs=*/ {m_coinbase_txns[2]}, /*output_values=*/ {10 * COIN});
+    pool.addUnchecked(entry.Fee(low_fee).FromTx(conflict_1));
+    const auto conflict_1_entry = pool.GetIter(conflict_1->GetHash()).value();
+
+    const auto conflict_2 = make_tx(/*inputs=*/ {m_coinbase_txns[3]}, /*output_values=*/ {10 * COIN});
+    pool.addUnchecked(entry.Fee(low_fee).FromTx(conflict_2));
+    const auto conflict_2_entry = pool.GetIter(conflict_2->GetHash()).value();
+
+    const auto conflict_3 = make_tx(/*inputs=*/ {m_coinbase_txns[4]}, /*output_values=*/ {10 * COIN});
+    pool.addUnchecked(entry.Fee(low_fee).FromTx(conflict_3));
+    const auto conflict_3_entry = pool.GetIter(conflict_3->GetHash()).value();
+
+    const auto err_string8{pool.CalculateFeerateDiagramsForRBF(/*replacement_fees=*/high_fee, /*replacement_vsize=*/low_size, {conflict_1_entry, conflict_2_entry, conflict_3_ent
ry}, {conflict_1_entry, conflict_2_entry, conflict_3_entry}, old_diagram, new_diagram)};
+
+    BOOST_CHECK(!err_string8.has_value());
+    BOOST_CHECK(old_diagram.size() == 4);
+    BOOST_CHECK(new_diagram.size() == 2);
+    old_diagram.clear();
+    new_diagram.clear();
+
+    // Add a child transaction to conflict_1 and make it cluster size 2
+    const auto conflict_1_child = make_tx(/*inputs=*/{conflict_1}, /*output_values=*/ {995 * CENT});
+    pool.addUnchecked(entry.Fee(low_fee).FromTx(conflict_1_child));
+    const auto conflict_1_child_entry = pool.GetIter(conflict_1_child->GetHash()).value();
+
+    const auto err_string9{pool.CalculateFeerateDiagramsForRBF(/*replacement_fees=*/high_fee, /*replacement_vsize=*/low_size, {conflict_1_entry, conflict_2_entry, conflict_3_ent
ry}, {conflict_1_entry, conflict_2_entry, conflict_3_entry, conflict_1_child_entry}, old_diagram, new_diagram)};
+
+    BOOST_CHECK(!err_string8.has_value());
+    BOOST_CHECK(old_diagram.size() == 4);
+    BOOST_CHECK(new_diagram.size() == 2);
+    old_diagram.clear();
+    new_diagram.clear();
+
+
+    // Add another descendant to conflict_1, making the cluster size > 2 should fail at this point.
+    const auto conflict_1_grand_child = make_tx(/*inputs=*/{conflict_1_child}, /*output_values=*/ {995 * CENT});
+    pool.addUnchecked(entry.Fee(high_fee).FromTx(conflict_1_grand_child));
+    const auto conflict_1_grand_child_entry = pool.GetIter(conflict_1_child->GetHash()).value();
+
+    const auto err_string10{pool.CalculateFeerateDiagramsForRBF(/*replacement_fees=*/high_fee, /*replacement_vsize=*/low_size, {conflict_1_entry, conflict_2_entry, conflict_3_en
try}, {conflict_1_entry, conflict_2_entry, conflict_3_entry, conflict_1_child_entry, conflict_1_grand_child_entry}, old_diagram, new_diagram)};
+
+    BOOST_CHECK(err_string10.has_value());
+    BOOST_CHECK(err_string10.value() == strprintf("%s has 2 descendants, max 1 allowed", conflict_1->GetHash().GetHex()));
+    old_diagram.clear();
+    new_diagram.clear();
 }
 
 BOOST_AUTO_TEST_CASE(feerate_diagram_utilities)
```
<details>

ismaelsadeeq reviewed (commented):
Thanks for adding the test!
re-ACK d57fbda350ee9051931d9a0dad4beb55f6b2e574

sdaftuar commented on src/txmempool.h:
```diff
@@ -733,6 +734,24 @@ class CTxMemPool
+     * @param[out] new_diagram        The feerate diagram of the relevant clusters after accepting the new tx
+     * @return An optional error string if the conflicts don't match a calculable topology
+     */
+    std::optional<std::string> CalculateFeerateDiagramsForRBF(CAmount replacement_fees, int64_t replacement_vsize, const setEntries& direct_conflicts, const setEntries& all_conflicts, std::vector<FeeFrac>& old_diagram, std::vector<FeeFrac>& new_diagram) EXCLUSIVE_LOCKS_REQUIRED(cs);
```
nit: Maybe update the comment to be more explicit that this function only works for the situation where (replacement_fees, replacement_size) corresponds to a transaction package that has no in-mempool dependences.

instagibbs replied:
updated to be more explicit on what those values represent

sdaftuar reviewed (commented):
Still reviewing all the tests -- leaving some nits that I've noticed so far.

In addition:
 * I commented [here](https://github.com/bitcoin/bitcoin/pull/29242#discussion_r1493358857) that I think we should drop the `Assume(size != 0 || fee == 0)` calls in feefrac.h, so that we don't have to police all the places we invoke operator- to make sure we can never accidentally violate that.
 * Another nit: in the commit message for commit 860823fb93212fa78ab928589bd403da462be222, it should say "FeeFrac" and not "FeeFrace"

Finally, as a note to other reviewers, I recently wrote up more explanation of the logic for using feerate diagram comparisons in our RBF policy here: https://delvingbitcoin.org/t/mempool-incentive-compatibility/553/1.

ismaelsadeeq commented on src/util/feefrac.cpp:
```diff
@@ -0,0 +1,93 @@
+// Copyright (c) The Bitcoin Core developers
```
Curious to know the full meaning of `FeeFrac` is it Fee Fraction?
Maybe Chunk is will be a better name?

instagibbs replied:
@sdaftuar going to hold off on this for now if that's ok

instagibbs commented on src/util/feefrac.cpp:
```diff
@@ -0,0 +1,93 @@
+
+    // Tail check at 0 feerate: Compare the remaining area. Use similar logic as in the loop above,
+    // except we use a horizontal line instead of AB, as no point B exists anymore.
+    const int long_side = next_index[1] != dias[1].size();
```
took in a modified form, left the direct exposure of `tail_feerate` as a YNGNI extension

instagibbs commented on src/policy/rbf.cpp:
```diff
@@ -181,3 +183,24 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
+                                                CAmount replacement_fees,
+                                                int64_t replacement_vsize)
+{
+    // Require that the replacement strictly improve the mempool's fee vs. size diagram.
```
changed to standard lingo

instagibbs commented on src/util/feefrac.cpp:
```diff
@@ -0,0 +1,93 @@
+        // let `P` be the next point on diagram unproc_side, and `A` and `B` the previous and next points
+        // on the other diagram. We want to know if P lies above or below the line AB. To determine this, we
+        // compute the direction coefficients of line AB and of line AP, and compare them. These
+        // direction coefficients are fee per size, and can thus be expressed as FeeFracs.
```
replaced, even though I thought it sounded pretty smart

sdaftuar commented on src/util/feefrac.cpp:
```diff
@@ -0,0 +1,89 @@
+        const FeeFrac& point_p = next_point(unproc_side);
+        const FeeFrac& point_a = prev_point(!unproc_side);
+
+        // Slope of AP can be negative, unlike AB
```
I think the slope of AB can be negative, because a user could use the `prioritisetransaction` to give a transaction a negative fee?

instagibbs replied:
one weird trick! will remove the comment

instagibbs replied:
un-resolved comment(oops), will add to follow-up

instagibbs replied:
opened in follow-up

sipa reviewed (commented):
Partial review.

hebasto commented:
Concept ACK on introducing feerate diagram checks for a subset of RBF transactions.

instagibbs commented on src/util/feefrac.cpp:
```diff
@@ -0,0 +1,89 @@
+    diagram.emplace_back(0, 0);
+    for (auto& chunk : chunks) {
+        FeeFrac& last = diagram.back();
+        diagram.emplace_back(last.fee+chunk.fee, last.size+chunk.size);
```
done

instagibbs commented on src/test/feefrac_tests.cpp:
```diff
@@ -0,0 +1,128 @@
+    chunks.push_back(oversized_2);
+
+    FastRandomContext rng{/*fDeterministic=*/true};
+    std::shuffle(chunks.begin(), chunks.end(), rng);
```
removed, since I stripped out the sorting function from `BuildDiagramFromUnsortedChunks`

ismaelsadeeq reviewed (approved):
Re-ACK 55eae36cfbd829eabc0772fc1b5312ec1f747e2c Reviewed [Diff](https://github.com/bitcoin/bitcoin/compare/3a2c66cc4f0afad9aff300110b18a39981055d63..55eae36cfbd829eabc0772fc1b5312ec1f747e2c)

Left some nits

instagibbs commented:
@ismaelsadeeq thanks, will fix up if I touch commits again

instagibbs commented on src/test/rbf_tests.cpp:
```diff
@@ -309,6 +309,162 @@ BOOST_FIXTURE_TEST_CASE(improves_feerate, TestChain100Setup)
+    old_diagram.clear();
+    new_diagram.clear();
+
+    // You can have more than two direct conflicts if the therea re multiple effected clusters, all of size 2 or less
```
done

instagibbs commented on src/test/feefrac_tests.cpp:
```diff
@@ -0,0 +1,124 @@
+
+    BOOST_CHECK(p1 == p1);
+    BOOST_CHECK(p1+p2 == sum);
+    BOOST_CHECK(p1-p2 == diff);
```
done

instagibbs commented on src/test/feefrac_tests.cpp:
```diff
@@ -0,0 +1,124 @@
+    BOOST_CHECK(oversized_1 != oversized_2);
+
+    // Tests paths that use double arithmetic
+    FeeFrac busted{((int64_t) INT32_MAX) + 1, INT32_MAX};
```
done

instagibbs commented:
took all suggestions and rebased

glozow commented on src/txmempool.h:
```diff
@@ -736,6 +737,28 @@ class CTxMemPool
+     * @param[in] direct_conflicts    All transactions that would be removed directly by
+     *                                having a conflicting input with a proposed transaction
+     * @param[in] all_conflicts       All transactions that would be removed
+     * @param[out] old_diagram        The feerate diagram of the relevant clusters before accepting the new tx
+     * @param[out] new_diagram        The feerate diagram of the relevant clusters after accepting the new tx
+     * @return An optional error string if the conflicts don't match a calculable topology
+     */
+    std::optional<std::string> CalculateFeerateDiagramsForRBF(CAmount replacement_fees, int64_t replacement_vsize, const setEntries& direct_conflicts, const setEntries& all_conflicts, std::vector<FeeFrac>& old_diagram, std::vector<FeeFrac>& new_diagram) EXCLUSIVE_LOCKS_REQUIRED(cs);
```
Did you consider using `util::Result` for this instead? Seems like you want `std::pair<std::vector<FeeFrac>, std::vector<FeeFrac>>` on success or a `std::string` on failure.

instagibbs replied:
done

glozow commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,160 @@
+};
+
+/** Takes the pre-computed and topologically-valid chunks and generates a fee diagram which starts at FeeFrac of (0, 0) */
+std::vector<FeeFrac> BuildDiagramFromChunks(const Span<const FeeFrac> chunks);
```
Would you consider creating an alias `using FeerateDiagram = std::vector<FeeFrac>`? Could be a bit easier to write/read.

instagibbs replied:
doesn't seem to save much typing/reading if we're using `Span`, leaving as is for now

glozow commented on src/txmempool.cpp:
```diff
@@ -1238,3 +1239,131 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+        // Additionally enforce that:
+        // If we have a child,  we are its only parent.
+        // If we have a parent, we are its only child.
+        if (has_descendant) {
+            const auto& our_child = direct_conflict->GetMemPoolChildrenConst().begin();
+            if (our_child->get().GetCountWithAncestors() > 2) {
+                return strprintf("%s is not the only parent of child %s",
+                                 txid_string, our_child->get().GetSharedTx()->GetHash().ToString());
+            }
+        } else if (has_ancestor) {
+            const auto& our_parent = direct_conflict->GetMemPoolParentsConst().begin();
+            if (our_parent->get().GetCountWithDescendants() > 2) {
+                return strprintf("%s is not the only child of parent %s",
+                                 txid_string, our_parent->get().GetSharedTx()->GetHash().ToString());
+            }
+        }
```
Nothing failed when I commented out this chunk, maybe need some 1p2c or 2p1c `CheckConflictTopology` tests?

instagibbs replied:
oops, bad test setup was missing this. Fixed and added string assertions to check what it's actually catching.

instagibbs replied:
I presume you mean the multi-child case didn't trigger failure when removed? multi-parent should

glozow reviewed (commented):
The code looks good, I'm still reading through the tests

instagibbs commented on src/test/rbf_tests.cpp:
```diff
@@ -269,6 +269,48 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
+    // It doesn't improve itself
+    const auto res1 = ImprovesFeerateDiagram(pool, {entry1}, {entry1, entry2}, tx1_fee + tx2_fee, tx1_size + tx2_size);
+    BOOST_CHECK(res1.has_value());
+    BOOST_CHECK(res1.value().first == DiagramCheckError::FAILURE);
```
seems I have to implement overloaded operator `std::ostream& operator<<(std::ostream& os, const FeeFrac& type)` to support it, and seems somewhat rare in the codebase. I can do it of course. Thoughts?

instagibbs replied:
ignoring for now

glozow commented on src/test/fuzz/rbf.cpp:
```diff
@@ -63,3 +81,98 @@ FUZZ_TARGET(rbf, .init = initialize_rbf)
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+    SetMockTime(ConsumeTime(fuzzed_data_provider));
+
+    std::optional<CMutableTransaction> mtx = ConsumeDeserializable<CMutableTransaction>(fuzzed_data_provider, TX_WITH_WITNESS);
+    if (!mtx) return;
```
Why make this ahead of time instead of just creating a new `CMutableTransaction` for each child?

instagibbs replied:
Aping the current rbf tests, probably. Picks a valid transaction structure, then sticks with it for the run. Suggestions for follow-up improvements maybe?

glozow commented on src/test/fuzz/feeratediagram.cpp:
```diff
@@ -0,0 +1,156 @@
+
+FUZZ_TARGET(build_and_compare_feerate_diagram)
+{
+    // Generate a random set of chunks
```
Maybe "same as above, but start with chunks instead of diagrams" ? Can any of the code maybe be helper funcitoned?

instagibbs replied:
removed the more limited fuzz test, as I don't think it adds additional coverage

glozow reviewed (commented):
Reviewed tests, just have some nits

instagibbs commented on src/test/rbf_tests.cpp:
```diff
@@ -309,6 +309,162 @@ BOOST_FIXTURE_TEST_CASE(improves_feerate, TestChain100Setup)
+    new_diagram.clear();
+
+    // Non-zero replacement fee/size
+    const auto err_string3{pool.CalculateFeerateDiagramsForRBF(/*replacement_fees=*/high_fee, /*replacement_vsize=*/low_size, {entry_low}, {entry_low}, old_diagram, new_diagram)};
```
an attempt was made!

instagibbs commented on src/txmempool.h:
```diff
@@ -736,6 +737,28 @@ class CTxMemPool
+     */
+    std::optional<std::string> CalculateFeerateDiagramsForRBF(CAmount replacement_fees, int64_t replacement_vsize, const setEntries& direct_conflicts, const setEntries& all_conflicts, std::vector<FeeFrac>& old_diagram, std::vector<FeeFrac>& new_diagram) EXCLUSIVE_LOCKS_REQUIRED(cs);
+
+    /* Check that all direct conflicts are in a cluster size of two or less. */
```
done

instagibbs commented on src/test/rbf_tests.cpp:
```diff
@@ -269,6 +269,48 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
+    const CAmount low_fee{CENT/100};
+    const CAmount normal_fee{CENT/10};
+
+    // One low feerate txn followed be normal feerate
```
taken

instagibbs commented on src/test/rbf_tests.cpp:
```diff
@@ -269,6 +269,48 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
+    // With one more satoshi it does
+    BOOST_CHECK(ImprovesFeerateDiagram(pool, {entry1}, {entry1, entry2}, tx1_fee + tx2_fee + 1, tx1_size + tx2_size) == std::nullopt);
+
+    // Make conflict un-calculable(for now)
```
we could expand acceptable topologies later. Took language.

instagibbs commented on src/test/fuzz/rbf.cpp:
```diff
@@ -63,3 +81,98 @@ FUZZ_TARGET(rbf, .init = initialize_rbf)
+    CTxMemPool pool{MemPoolOptionsForTest(g_setup->m_node)};
+
+    // Add a bunch of parent-child pairs to the mempool, and remember them.
+    std::vector<CTransaction> txs;
```
done

instagibbs commented on src/test/fuzz/rbf.cpp:
```diff
@@ -23,12 +23,30 @@ namespace {
+
+    // Create a fixed set of unique "UTXOs" to source parents from
+    // to avoid fuzzer giving circular references
+    for (int i = 0; i < NUM_ITERS * 2; ++i) {
```
removed

instagibbs commented on src/test/fuzz/rbf.cpp:
```diff
@@ -63,3 +81,98 @@ FUZZ_TARGET(rbf, .init = initialize_rbf)
+
+    // If internals report error, wrapper should too
+    auto err_tuple{ImprovesFeerateDiagram(pool, direct_conflicts, all_conflicts, replacement_fees, replacement_vsize)};
+    if (err_string.has_value()) assert(err_tuple.has_value());
```
I guess the question is if unique mempools per run are worth the overhead. I think it's pretty fast still?

instagibbs commented:
I believe I addressed all comments.

dergoegge commented:
Did some more fuzzing. No crashes. Coverage: http://bitcoind-fuzz.dergoegge.de/instagibbs/bitcoin/2024-01-diagram-checks/coverage/

instagibbs commented on src/policy/rbf.cpp:
```diff
@@ -181,3 +183,24 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
+
+    const auto diagram_results{pool.CalculateFeerateDiagramsForRBF(replacement_fees, replacement_vsize, direct_conflicts, all_conflicts)};
+
+    if (!diagram_results.has_value()) {
```
`util::Result` lets you pass an error string in lieu of a `std::nullopt`, and based on the docs it seems I'm using this right. @glozow ?


```
//! `std::optional<T>` can be updated to return `util::Result<T>` and return
//! error strings usually just replacing `return std::nullopt;` with `return
//! util::Error{error_string};`.
```

instagibbs replied:
this is what it's returning instead of a `std::nullopt`: https://github.com/bitcoin/bitcoin/pull/29242/files#diff-c065d4cd2398ad0dbcef393c5dfc53f465bf44723348892395fffd2fb3bac522R1290

(we might be talking past each other)

instagibbs commented:
@sipa to keep things moving (muh package rbf) I'll postpone/ promise to review your PR for that commit as a follow-up.

instagibbs commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,160 @@
+};
+
+/** Takes the pre-computed and topologically-valid chunks and generates a fee diagram which starts at FeeFrac of (0, 0) */
+std::vector<FeeFrac> BuildDiagramFromChunks(const Span<const FeeFrac> chunks);
```
I probably should just be passing a reference here anyways? Changed to pass by reference

instagibbs replied:
got it, double checked, taking your suggestion to remove const from declaration instead

instagibbs commented on src/txmempool.cpp:
```diff
@@ -1238,3 +1239,131 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+    }
+
+    // No topology restrictions post-chunking; sort
+    std::sort(old_chunks.begin(), old_chunks.end(), [](const FeeFrac& a, const FeeFrac& b) { return a > b; });
```
done

sipa reviewed (commented):
utACK 72959867784098137a50c34f86deca8235eef4f8

Some non-blocking nits:

glozow commented on src/test/rbf_tests.cpp:
```diff
@@ -294,6 +294,48 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
+
+    // With one more satoshi it does
+    BOOST_CHECK(ImprovesFeerateDiagram(pool, {entry1}, {entry1, entry2}, tx1_fee + tx2_fee + 1, tx1_size + tx2_size) == std::nullopt);
+
```
```suggestion
    // With one less vB it does
    BOOST_CHECK(ImprovesFeerateDiagram(pool, {entry1}, {entry1, entry2}, tx1_fee + tx2_fee, tx1_size + tx2_size - 1) == std::nullopt);
```

instagibbs replied:
opened in follow-up

glozow commented on src/test/fuzz/rbf.cpp:
```diff
@@ -63,3 +81,98 @@ FUZZ_TARGET(rbf, .init = initialize_rbf)
+
+    // If internals report error, wrapper should too
+    auto err_tuple{ImprovesFeerateDiagram(pool, direct_conflicts, all_conflicts, replacement_fees, replacement_vsize)};
+    if (!calc_results.has_value()) assert(err_tuple.has_value());
```
I think we can be a bit more specific than this, i.e. if `!calc_results.has_value()` then `err_tuple.value().first == DiagramCheckError::UNCALCULABLE`

instagibbs replied:
opened in follow-up

glozow commented on src/test/fuzz/rbf.cpp:
```diff
@@ -63,3 +81,98 @@ FUZZ_TARGET(rbf, .init = initialize_rbf)
+        // The total fee of the new diagram should be the total fee of the old
+        // diagram - replaced_fee + replacement_fees
+        assert(calc_results->first.back().fee - replaced_fee + replacement_fees == calc_results->second.back().fee);
+        assert(calc_results->first.back().size - replaced_size + replacement_vsize == calc_results->second.back().size);
```
This test for `ImprovesFeerateDiagram` doesn't seem very strict beyond checking NEW = OLD - CON + replacement and that internal errors = wrapper errors. For example, if I swap `!is_gt` for a `!is_eq` in `ImprovesFeerateDiagram`, this fuzzer is still happy with everything. There is coverage for that in other places, but it seems like we could be writing stronger assertions?

Can we perhaps add a check that the result is consistent with what `CompareFeerateDiagram` says when comparing the diagrams returned in `calc_results`? Or some kind of sanity check that when the new diagram ends with lower fees, we should be getting `DiagramCheckError::FAILURE`?

instagibbs replied:
I added a couple more checks where I could clearly add them in followup

glozow reviewed (commented):
code review ACK 72959867784098137a50c34f86deca8235eef4f8

Various non-blocking comments about the tests, can be saved for a followup.

instagibbs commented:
Thanks for the review!

As long as it's nits only, I will compile a list of :+1: I will accomplish on a follow-up PR.

murchandamus commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,160 @@
+#endif
+
+    int64_t fee;
+    int32_t size;
```
In "Add FeeFrac utils" (ce8e22542ed0b4fa5794d3203207146418d59473):

Nit: I noticed that in `CalculateFeerateDiagramsForRBF(…)` you explicitly use `vsize`, but here you use `size`. Perhaps `vsize` should be used consistently? Using `vsize` for the FeeFrac member would also have the added benefit that the member name doesn’t collide with the use of `size` referring to the count of objects in the diagram.

```suggestion
    int32_t vsize;
```

instagibbs replied:
> Should they perhaps just be called numerator and denominator?

I'll mull this over a bit

murchandamus replied:
If it were `weight` that would also differentiate well.

instagibbs replied:
I think that would be even more confusing; if we're going to change it we need it to be unit-neutral (punting for this PR)

murchandamus commented on src/test/fuzz/feeratediagram.cpp:
```diff
@@ -0,0 +1,119 @@
+    unsigned not_above = 0;
+    unsigned not_below = diagram.size() - 1;
+    // If outside the range of diagram, extend begin/end.
+    if (size < diagram[not_above].size) return {diagram[not_above].fee, 1};
```
In "fuzz: fuzz diagram creation and comparison" (4d6528a3d6bf3821c216c68f99170e2faab5d63c):

Nit: The use of `vsize` throughout would help differentiate the mixed occurrences of `diagram.size()` and `diagram[i].size`

instagibbs replied:
leaving as is

murchandamus commented on src/test/rbf_tests.cpp:
```diff
@@ -334,6 +334,164 @@ BOOST_FIXTURE_TEST_CASE(improves_feerate, TestChain100Setup)
+    BOOST_CHECK(new_diagram[0] == FeeFrac(0, 0));
+    BOOST_CHECK(new_diagram[1] == FeeFrac(high_fee, low_size_2));
+
+    // You can have more than two direct conflicts if the there are multiple effected clusters, all of size 2 or less
```
Nit: In "Unit tests for CalculateFeerateDiagramsForRBF" (72959867784098137a50c34f86deca8235eef4f8):

```suggestion
    // You can have more than two direct conflicts if the there are multiple affected clusters, all of size 2 or less
```

instagibbs replied:
opened in follow-up

murchandamus commented on src/test/rbf_tests.cpp:
```diff
@@ -334,6 +334,164 @@ BOOST_FIXTURE_TEST_CASE(improves_feerate, TestChain100Setup)
+    const CAmount normal_fee{CENT/10};
+    const CAmount high_fee{CENT};
+
+    // low -> high -> medium fee transactions that would result in two chunks together
```
In "Unit tests for CalculateFeerateDiagramsForRBF" (72959867784098137a50c34f86deca8235eef4f8):

Nit: this comment appears to make the unstated assumption that the three transactions are of similar vsize

murchandamus reviewed (commented):
utACK 72959867784098137a50c34f86deca8235eef4f8

I also only have non-blocking nits

instagibbs commented on src/txmempool.cpp:
```diff
@@ -1238,3 +1239,131 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+        }
+    }
+
+    // No topology restrictions post-chunking; sort
```
> If that’s what is being alluded to here, perhaps you could elaborate in the comment if you touch this again.

Will elaborate a bit on the chunk construction comments in a follow-up.

> I think that's generally an improvement, actually, because size shouldn't matter here.

Agreed, at a minimum I'd change to `individual >> package` to get rid of the tie-breaking behavior which is confusing to have here. Will take on a chosen change in follow-up.

instagibbs replied:
opened in follow-up with `>>` as noted

ismaelsadeeq reviewed (commented):
Re-ACK https://github.com/bitcoin/bitcoin/commit/72959867784098137a50c34f86deca8235eef4f8

willcl-ark reviewed (approved):
crACK 72959867784098137a50c34f86deca8235eef4f8

This looks good on a first pass of the code, and was a good introduction (to me) for how we are approaching implementation of the Diagrammatic fee rate checks, which matches my current understanding of the proposed concept.

Left only two small non-blocking nits.

In addition, while checking each commit I did notice that clang-format could be run on most (all?) of them too if wanted (but I'd guess it's not wanted at this stage of review 😄 ).

instagibbs commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,160 @@
+ * A FeeFrac is considered "better" if it sorts after another, by this ordering. All standard
+ * comparison operators (<=>, ==, !=, >, <, >=, <=) respect this ordering.
+ *
+ * The CompareFeeFrac, and >> and << operators only compare feerate and treat equal feerate but
```
opened in follow-up

instagibbs commented on src/policy/rbf.cpp:
```diff
@@ -181,3 +183,24 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
+                                                CAmount replacement_fees,
+                                                int64_t replacement_vsize)
+{
+    // Require that the replacement strictly improve the mempool's feerate diagram.
```
opened in follow-up

instagibbs commented on src/test/rbf_tests.cpp:
```diff
@@ -215,15 +248,119 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
+
+    BOOST_CHECK(std::is_lt(CompareFeerateDiagram(old_diagram, new_diagram)));
+
+    // Incomparable diagrams
+    old_diagram = {FeeFrac{0, 0}, FeeFrac{950, 300}, FeeFrac{1050, 400}};
+    new_diagram = {FeeFrac{0, 0}, FeeFrac{1000, 300}, FeeFrac{1000, 400}};
+
+    BOOST_CHECK(CompareFeerateDiagram(old_diagram, new_diagram) == std::partial_ordering::unordered);
+
+    // Strictly better but smaller size.
+    old_diagram = {FeeFrac{0, 0}, FeeFrac{950, 300}, FeeFrac{1050, 400}};
+    new_diagram = {FeeFrac{0, 0}, FeeFrac{1100, 300}};
+
+    BOOST_CHECK(std::is_lt(CompareFeerateDiagram(old_diagram, new_diagram)));
+
+    // New diagram is strictly better due to the first chunk, even though
+    // second chunk contributes no fees
+    old_diagram = {FeeFrac{0, 0}, FeeFrac{950, 300}, FeeFrac{1050, 400}};
+    new_diagram = {FeeFrac{0, 0}, FeeFrac{1100, 100}, FeeFrac{1100, 200}};
+
+    BOOST_CHECK(std::is_lt(CompareFeerateDiagram(old_diagram, new_diagram)));
+
+    // Feerate of first new chunk is better with, but second chunk is worse
+    old_diagram = {FeeFrac{0, 0}, FeeFrac{950, 300}, FeeFrac{1050, 400}};
+    new_diagram = {FeeFrac{0, 0}, FeeFrac{750, 100}, FeeFrac{999, 350}, FeeFrac{1150, 1000}};
+
+    BOOST_CHECK(CompareFeerateDiagram(old_diagram, new_diagram) == std::partial_ordering::unordered);
+
+    // If we make the second chunk slightly better, the new diagram now wins.
+    old_diagram = {FeeFrac{0, 0}, FeeFrac{950, 300}, FeeFrac{1050, 400}};
+    new_diagram = {FeeFrac{0, 0}, FeeFrac{750, 100}, FeeFrac{1000, 350}, FeeFrac{1150, 500}};
+
+    BOOST_CHECK(std::is_lt(CompareFeerateDiagram(old_diagram, new_diagram)));
+
+    // Identical diagrams, cannot be strictly better
+    old_diagram = {FeeFrac{0, 0}, FeeFrac{950, 300}, FeeFrac{1050, 400}};
+    new_diagram = {FeeFrac{0, 0}, FeeFrac{950, 300}, FeeFrac{1050, 400}};
+
+    BOOST_CHECK(std::is_eq(CompareFeerateDiagram(old_diagram, new_diagram)));
+
```
opened in follow-up

instagibbs commented on src/test/rbf_tests.cpp:
```diff
@@ -215,15 +248,119 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
+    // Tests for CheckConflictTopology
+
+    // Tx4 has 23 descendants
+    BOOST_CHECK(pool.CheckConflictTopology(set_34_cpfp).has_value());
```
opened in follow-up

instagibbs commented on src/policy/rbf.cpp:
```diff
@@ -181,3 +183,24 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
+                                                int64_t replacement_vsize)
+{
+    // Require that the replacement strictly improve the mempool's feerate diagram.
+    std::vector<FeeFrac> old_diagram, new_diagram;
```
opened in follow-up

instagibbs commented on src/test/rbf_tests.cpp:
```diff
@@ -334,6 +334,164 @@ BOOST_FIXTURE_TEST_CASE(improves_feerate, TestChain100Setup)
+    const auto entry_low = pool.GetIter(low_tx->GetHash()).value();
+    const auto low_size = entry_low->GetTxSize();
+
+    std::vector<FeeFrac> old_diagram, new_diagram;
+
+    // Replacement of size 1
+    const auto replace_one{pool.CalculateFeerateDiagramsForRBF(/*replacement_fees=*/0, /*replacement_vsize=*/1, {entry_low}, {entry_low})};
+    BOOST_CHECK(replace_one.has_value());
+    old_diagram = replace_one->first;
+    new_diagram = replace_one->second;
+    BOOST_CHECK(old_diagram.size() == 2);
+    BOOST_CHECK(new_diagram.size() == 2);
+    BOOST_CHECK(old_diagram[0] == FeeFrac(0, 0));
+    BOOST_CHECK(old_diagram[1] == FeeFrac(low_fee, low_size));
+    BOOST_CHECK(new_diagram[0] == FeeFrac(0, 0));
+    BOOST_CHECK(new_diagram[1] == FeeFrac(0, 1));
```
opened in follow-up

instagibbs commented on src/test/fuzz/rbf.cpp:
```diff
@@ -63,3 +81,98 @@ FUZZ_TARGET(rbf, .init = initialize_rbf)
+        parent->vin[0].prevout = g_outpoints[iter++];
+
+        mempool_txs.emplace_back(*parent);
+        pool.addUnchecked(ConsumeTxMemPoolEntry(fuzzed_data_provider, mempool_txs.back()));
```
opened in follow-up

instagibbs commented on src/test/rbf_tests.cpp:
```diff
@@ -294,6 +294,48 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
+    const CAmount normal_fee{CENT/10};
+
+    // low feerate parent with normal feerate child
+    const auto tx1 = make_tx(/*inputs=*/ {m_coinbase_txns[0]}, /*output_values=*/ {10 * COIN});
+    pool.addUnchecked(entry.Fee(low_fee).FromTx(tx1));
```
opened in follow-up

sdaftuar commented on src/util/feefrac.cpp:
```diff
@@ -0,0 +1,86 @@
+        if (std::is_gt(cmp)) better_somewhere[unproc_side] = true;
+        if (std::is_lt(cmp)) better_somewhere[!unproc_side] = true;
+        ++next_index[unproc_side];
+    } while(true);
```
Just noticed that I think we could change this to short circuit in the case where we've detected that both diagrams are "better somewhere" and hence incomparable, eg:

```
} while (!(better_somewhere[0] && better_somewhere[1]));
```

instagibbs replied:
I think slightly better is to pull the check a few lines below into the loop, just after updating the `better_somewhere` array:

```
if (better_somewhere[0] && better_somewhere[1]) return std::partial_ordering::unordered;
```

?

instagibbs replied:
e.g., https://github.com/bitcoin/bitcoin/pull/29724/commits/a9d42b9aa579f54922ffd17fdeb61e704539b92c

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -1238,3 +1239,131 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+
+    // old diagram will consist of each element of all_conflicts either at
+    // its own feerate (followed by any descendant at its own feerate) or as a
+    // single chunk at its descendant's ancestor feerate.
```
I think this comment is wrong -- it should say something like:

```
// old diagram will consist of the ancestors and descendants of each element of 
// all_conflicts.  every such transaction will either be at its own feerate (followed 
// by any descendant at its own feerate), or as a single chunk at the descendant's 
// ancestor feerate.
```

sdaftuar reviewed (approved):
ACK 72959867784098137a50c34f86deca8235eef4f8

Left some nits that can be addressed (or not) in a followup.
